src/main/scala/me/jeffmay/neo4j/client/ws/WSNeo4jClient.scala

```diff
@@ -0,0 +1,208 @@
+    // TODO: Allow configuration per-request
+    val jsonBody = Json.toJson(RawNeo4jCypherRequest(statements, includeStats = Some(config.includeStatsByDefault)))
+    val request = http("/db/data/transaction/commit")
+    whenInitialized {
+      request.postAndCheckStatus(Some(jsonBody)).flatMap { resp =>
+        // TODO: Handle version / json format errors with common recovery code
+        val respBody = resp.json.as[RawTxnResponse]
+        if (respBody.isSuccessful) {
+          Future.successful(respBody.asCommitTxnResponse.get)
+        }
+        else {
+          Future.failed(statusCodeException(request, Some(jsonBody), resp, respBody.errors))
+        }
+      }
+    }
+  }
+
+  override def commitTxn(ref: TxnRef, statements: Seq[CypherStatement]): Future[CommittedTxnResponse] = {
+    // TODO: Allow configuration per-request
+    val jsonBody = Json.toJson(RawNeo4jCypherRequest(statements, includeStats = Some(config.includeStatsByDefault)))
+    val request = http("/db/data/transaction")
+    whenInitialized {
+      request.postAndCheckStatus(Some(jsonBody)).flatMap { resp =>
+        // TODO: Handle version / json format errors with common recovery code
+        val respBody = resp.json.as[RawTxnResponse]
+        if (respBody.isSuccessful) {
+          Future.successful(respBody.asCommitTxnResponse.get)
+        }
+        else {
+          Future.failed(statusCodeException(request, Some(jsonBody), resp, respBody.errors))
+        }
+      }
+    }
+  }
+}
+
+object WSNeo4jClient {
+
+  // TODO: Replace this with an actor to manage throttling / circuit breaker / backoff retry on unauthorized or connection issues
+
+  @volatile private var lastSuccessfulRequestWithCreds: Option[Neo4jBasicAuth] = None
+
+  // this is a very naive implementation of recovery that should evolve into a more robust / non-blocking solution
+  private def changePasswordWithCredentials(client: Neo4jClient, creds: Neo4jBasicAuth): Unit = {
+    blocking {
+      synchronized {
+        // TODO: Is username needed for this?
+        Await.result(client.changePassword(creds.password), 1.second)
+        // If no exception is thrown, let's update the creds
+        lastSuccessfulRequestWithCreds = Some(creds)
+      }
+    }
+  }
+
+  private def withVerifiedCredentials[T](client: Neo4jClient, creds: Neo4jBasicAuth)(f: => Future[T])
+    (implicit ec: ExecutionContext): Future[T] = {
+    f.recoverWith {
+      case resp: UnexpectedStatusException if resp.responseStatus == Status.UNAUTHORIZED =>
+        WSNeo4jClient.synchronized {
+          val currentCreds = lastSuccessfulRequestWithCreds
+          lastSuccessfulRequestWithCreds = None
+          if (!(currentCreds contains creds)) {
+            Future(changePasswordWithCredentials(client, creds)).flatMap(_ => f)
+          }
+          else {
+            Future.failed(resp)
+          }
+        }
+    }
+  }
```

Comment:
Basically, this code just makes sure to synchronize all recover paths through the singleton `WSNeo4jClient`. This blocks all threads until the credentials are good again. This isn't a terrible way to solve this particular issue because you only ever run into this situation on startup when the database needs to be reloaded.

However, if the server is getting hammered on startup, then blocking all threads until the password is changed could by a risky point of failure. Ideally, we could use a non-blocking approach that allows the server to queue up requests with small objects (ie. not threads) and then handle them quickly. If we are getting 10,000 requests / second, then that would be 10 requests / ms, and if it takes 10ms to change the password, that would be 100 threads that are blocked. After that, the Neo4j server would not have time to warm up the cache before being bombarded with a shit ton of queries.

It isn't a show stopper for a small project, but if we wanted to support building the graph at boot, it would be a weak point for using this driver.

I'm not going to worry about it for the V1 launch of this project unless it becomes higher priority, but I just wanted to call it out.